neural_lam/datastore/base.py

```diff
@@ -0,0 +1,371 @@
+# Standard library
+import abc
+import collections
+import dataclasses
+from pathlib import Path
+from typing import List, Union
+
+# Third-party
+import cartopy.crs as ccrs
+import numpy as np
+import xarray as xr
+
+
+class BaseDatastore(abc.ABC):
+    """
+    Base class for weather data used in the neural- lam package. A datastore
+    defines the interface for accessing weather data by providing methods to
+    access the data in a processed format that can be used for training and
+    evaluation of neural networks.
+
+    NOTE: All methods return either primitive types, `numpy.ndarray`,
+    `xarray.DataArray` or `xarray.Dataset` objects, not `pytorch.Tensor`
+    objects. Conversion to `pytorch.Tensor` objects should be done in the
+    `weather_dataset.WeatherDataset` class (which inherits from
+    `torch.utils.data.Dataset` and uses the datastore to access the data).
+
+    # Forecast vs analysis data
+    If the datastore is used represent forecast rather than analysis data, then
+    the `is_forecast` attribute should be set to True, and returned data from
+    `get_dataarray` is assumed to have `analysis_time` and `forecast_time`
+    dimensions (rather than just `time`).
+
+    # Ensemble vs deterministic data
+    If the datastore is used to represent ensemble data, then the `is_ensemble`
+    attribute should be set to True, and returned data from `get_dataarray` is
+    assumed to have an `ensemble_member` dimension.
+
+    """
+
+    is_ensemble: bool = False
```

Comment:
(I changed this to a code comment so it can be resolved when we are done discussing it)
## Regarding allowing for ensemble data in the datastore

I have realized there are a few things making the training on ensemble data tricky. This is now handled by having a property `is_ensemble=True` in the datastore, that other methods then can check.

1. Random sampling of ensemble members: https://github.com/leifdenby/neural-lam/blob/1f661c6e008d69b7c0daf0fff8a2135b62b0848d/neural_lam/weather_dataset.py#L172-L178. I recognize that we have to choose which member a specific `idx` corresponds to, but I think this is a bad idea. This would mean that you get different members when running evaluation on e.g. the test set, meaning that your evaluation metrics will depend heavily on the random seed used.
2. https://github.com/leifdenby/neural-lam/blob/1f661c6e008d69b7c0daf0fff8a2135b62b0848d/neural_lam/datastore/npyfiles/store.py#L162 A *real* headache here is that the MEPS data contains a different number of ensemble members for the different analysis times :fearful:. So specifying a fixed number of members in the datastore will not really work. This also means that it is simply not possible to have a `ensemble_member` dimension in the returned `xr.Dataarray` (at least with out some very finicky masking of missing members).

A solution to all of this would be to just treat each ensemble member as a separate independent sample (this is the way it is currently done). However, that becomes very tricky with this `xr.Dataarray` setup, as multiple members will have the same analysis time. I suppose a way to get around it would be to have a version of a datastore that does not have an analysis time dimension, but rather just a sample index dimensions. But then you don't know what time your samples are from.

Overall it seems to me that allowing for ensemble training data at this point is more work than it is worth. If this becomes relevant I think we can figure it out later. For now I am leaning towards just getting rid of the `is_ensemble` option and restricting the MEPS numpy datastore to just use the control member (`mbr0`). That one is always available.

Reply:
> 1. Random sampling of ensemble members: https://github.com/leifdenby/neural-lam/blob/1f661c6e008d69b7c0daf0fff8a2135b62b0848d/neural_lam/weather_dataset.py#L172-L178. I recognize that we have to choose which member a specific `idx` corresponds to, but I think this is a bad idea. This would mean that you get different members when running evaluation on e.g. the test set, meaning that your evaluation metrics will depend heavily on the random seed used.

Very good point @joeloskarsson, I hadn't thought of this. Ok, we can't do random sampling. 

> 2\. A _real_ headache here is that the MEPS data contains a different number of ensemble members for the different analysis times 😨.

OH! I had no idea... Yes that makes sampling even more tricky. I would definitely prefer to keep the time since that makes all the other functionality much easier. For example if we decided to combine both time and ensemble member into a single index then we would need to add logic somewhere that the sampling doesn't go across different ensemble members. Might it be worth considering subsets of MEPS data which have the same number as ensemble members as logically separate datasets (I assume the number of members increased over time)? One could then build pytorch-datasets for each (using the existing functionality in this PR) and then make a sort of composite pytorch dataset that is comprised of a multiple WeatherDatasets. That class would then simply use the __len__ method on each child dataset to work out the total size and proxy the __getitem__ call to the correct pytorch dataset. Again, this is not for now, but just an idea.

> For now I am leaning towards just getting rid of the `is_ensemble` option and restricting the MEPS numpy datastore to just use the control member (`mbr0`).

Ok, for now I would prefer to keep the code as it is, but we only sample the first ensemble member and issue a warning to the user that sampling from multiple hasn't been implemented yet. If the number of ensemble members is constant across time it would be quite easy to extend the current sampling to include them all, but we can leave that for now.

Reply:
> (I assume the number of members increased over time)

Sadly this is not even the case :sweat: It is just that at some seemingly random times there are one or more ensemble members missing, due to operational issues. But the workaround you describe could still work I think. But as you say, not worth doing now, but a potential fix if this becomes relevant for the future.

> Ok, for now I would prefer to keep the code as it is, but we only sample the first ensemble member and issue a warning to the user that sampling from multiple hasn't been implemented yet. If the number of ensemble members is constant across time it would be quite easy to extend the current sampling to include them all, but we can leave that for now.

I think that sounds good! It makes sense that if you have an ensemble dataset you should make sure to have all members at all times. The MEPS case should not set the precedent.

Reply:
Ok, I've implemented only using the first ensemble member for now. Thanks again